### PR TITLE
Fix span mapping in completion commit

### PR DESCRIPTION
**Pull request by rchande**

### CODE REVIEW

**Comment by rchande** [comment]
retest vsi please

**Comment by rchande** [comment]
retest vsi please

**Comment by rchande** [comment]
retest vsi please

**Comment by rchande** [comment]
retest vsi please

**Comment by rchande** [comment]
retest vsi please

**Comment by rchande** [comment]
retest vsi please

**Comment by rchande** [comment]
The `CsharpErrorLog` VSI tests seem to fail, however, they don't even type so they are unrelated to this change.
Similarly, the ETA `Peek` test fails, but it also doesn't type.

**Comment by rchande** [comment]
@balajikris: I'm seeing BasicVenus fail with snippet issues. Is that a known issue?

**Comment by rchande** [comment]
Tagging @dotnet/roslyn-ide @CyrusNajmabadi @paulvanbrenk for review.

**Comment by balajikris** [comment]
@rchande : something doesnt sound right there. I thought snippets scenario in BasicVenus was disabled - see https://github.com/dotnet/roslyn/issues/12288

**Comment by rchande** [comment]
Ah, @balajikris I bet that change hasn't been ported to the microupdate branch.

**Comment by balajikris** [comment]
nice. that should be it. so, its a definite failure and will fail again. can you ignore that failure case?

**Comment by paulvanbrenk** [comment]
See no changes that would affect TypeScript, so LGTM

**Comment by rchande** [comment]
@balajikris Yeah, all the tests that are failing for me were disabled by Tanner in bf59b68123ebac2642eeaefb3873cfcf3443950c which is not in microupdate.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs [hunk 1]
```diff
@@ -95,23 +94,39 @@ private void Commit(PresentationItem item, Model model, char? commitChar, Cancel
...
+                        var translatedOriginalSpanInView = originalSpanInView.TranslateTo(TextView.TextBuffer.CurrentSnapshot, SpanTrackingMode.EdgeInclusive);
+
+                        textChangesInCurrentViewSnapshot.Add(new TextChange(translatedOriginalSpanInView.Span.ToTextSpan(), triggerSnapshotTextChange.NewText));
+                    }
+
+                    // Note: we currently create the edit on the textview's buffer.  This is 
+                    // necessary as our own 
+                    using (var textEdit = TextView.TextBuffer.CreateEdit(editOptions, reiteratedVersionNumber: null, editTag: null))
+                    {
+                        for (int iChange = 0; iChange < textChangesAgainstTriggerSnapshot.Length; iChange++)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: i don't like that we're using two different arrays for length and for indexing.  can you only reference one of the arrays here?

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs [hunk 2]
```diff
@@ -95,23 +94,39 @@ private void Commit(PresentationItem item, Model model, char? commitChar, Cancel
...
+                    foreach (var triggerSnapshotTextChange in textChangesAgainstTriggerSnapshot)
                     {
-                        for (int iChange = 0; iChange < textChanges.Length; iChange++)
+                        var currentSpanInBuffer = new SnapshotSpan(
+                            model.TriggerSnapshot,
+                            triggerSnapshotTextChange.Span.ToSpan());
+
+                        // Try mapping the item span against the trigger snapshot to the ViewBuffer's trigger snapshot.
+                        // Then map that forward to the ViewBuffer's current snapshot.
+                        var originalSpanInView = model.GetViewBufferSpan(triggerSnapshotTextChange.Span).TextSpan.ToSnapshotSpan(model.TriggerSnapshot);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: This line looks wrong to me.  We have a span relative to the snapshot for the subject buffer .  We then map that to the view buffer's snapshot.    BUt then we just ask for that span against the trigger snapshot... 

The trigger snapshot corresponds to the subject buffer.  But the view buffer is totallly different...

**Comment by Pilchie** [comment]
@tannergooding has a PR out to also disabled them here - should hopefully be merged soon.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs [hunk 3]
```diff
@@ -95,23 +94,39 @@ private void Commit(PresentationItem item, Model model, char? commitChar, Cancel
...
+                    foreach (var triggerSnapshotTextChange in textChangesInTriggerSnapshot)
                     {
-                        for (int iChange = 0; iChange < textChanges.Length; iChange++)
+                        var currentSpanInBuffer = new SnapshotSpan(
+                            model.TriggerSnapshot,
+                            triggerSnapshotTextChange.Span.ToSpan());
+
+                        // Try mapping the item span against the trigger snapshot to the ViewBuffer's trigger snapshot.
+                        // Then map that forward to the ViewBuffer's current snapshot.
+                        var originalSpanInView = model.GetViewBufferSpan(triggerSnapshotTextChange.Span).TextSpan.ToSnapshotSpan(model.ViewTriggerSnapshot);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: why is the latter part of this expression necessary?  You alreasy have the viewbuffer span... but then you take it, unpack it, and then repack it.  I'm not really getting why.

**Comment by rchande** [comment]
Need test:

- [ ] Debugger intellisense
- [ ] Workflow

- [ ] Diff view

**Comment by rchande** [comment]
retest vsi please

**Comment by rchande** [comment]
retest vsi please

**Comment by rchande** [comment]
I'd like to get this merged today. Any other feedback?

**Comment by Pilchie** [comment]
Thanks @rchande  :+1:

**Pull request merged**